pennylane/qnn.py

```diff
@@ -0,0 +1,205 @@
+                "QNode must include an argument with name {} for inputting data".format(INPUT_ARG)
+            )
+        if INPUT_ARG in set(weight_shapes.keys()):
+            raise ValueError(
+                "{} argument should not have its dimension specified in "
+                "weight_shapes".format(INPUT_ARG)
+            )
+        if set(weight_shapes.keys()) | {INPUT_ARG} != set(self.sig.keys()):
+            raise ValueError("Must specify a shape for every non-input parameter in the QNode")
+        if qnode.func.var_pos:
+            raise TypeError("Cannot have a variable number of positional arguments")
+        if qnode.func.var_keyword:
+            raise TypeError("Cannot have a variable number of keyword arguments")
+
+        self.qnode = to_tf(qnode, dtype=tf.keras.backend.floatx())
+        self.weight_shapes = {
+            weight: (tuple(size) if isinstance(size, Iterable) else (size,))
+            for weight, size in weight_shapes.items()
+        }
+        self.output_dim = output_dim[0] if isinstance(output_dim, Iterable) else output_dim
+
+        defaults = {
+            name for name, sig in self.sig.items() if sig.par.default != inspect.Parameter.empty
+        }
+        self.input_is_default = INPUT_ARG in defaults
+        if defaults - {INPUT_ARG} != set():
+            raise TypeError("Only the argument {} is permitted to have a default".format(INPUT_ARG))
+
+        self.weight_specs = weight_specs if weight_specs is not None else {}
+
+        self.qnode_weights = {}
+
+        super(KerasLayer, self).__init__(dynamic=True, **kwargs)
+
+    def build(self, input_shape):
+        """Initializes the QNode weights.
+
+        Args:
+            input_shape (tuple or tf.TensorShape): shape of input data
+        """
+        for weight, size in self.weight_shapes.items():
+            spec = self.weight_specs.get(weight, {})
+            self.qnode_weights[weight] = self.add_weight(name=weight, shape=size, **spec)
+
+        super(KerasLayer, self).build(input_shape)
+
+    def call(self, inputs):
+        """Evaluates the QNode on input data using the initialized weights.
+
+        Args:
+            inputs (tensor): data to be processed
+
+        Returns:
+            tensor: output data
+        """
+        outputs = []
+        for x in inputs:
+            qnode = self.qnode
+            for arg in self.sig:
+                if arg is not INPUT_ARG:
+                    w = self.qnode_weights[arg]
+                    if w.shape == (1,):
+                        qnode = functools.partial(qnode, w[0])
+                    else:
+                        qnode = functools.partial(qnode, w)
+                else:
+                    if self.input_is_default:
+                        qnode = functools.partial(qnode, **{INPUT_ARG: x})
+                    else:
+                        qnode = functools.partial(qnode, x)
+            outputs.append(qnode())
+
+        return tf.stack(outputs)
```

Comment:
I'm not quite clear what this method is doing, could you explain the logic/reasoning? Might also be good to comment :)

Reply:
That's perfect, thanks Tom! Would suggest putting the above explanation in comments (either line or block comments) in the code itself.

Reply:
Oops I forgot to add the comments! But have done this now.